Design: Debugger Support for Multi-Module Enclaves (#3838)
openenclave/openenclave: docs/DesignDocs/DebuggerSupportForMultiModuleEnclaves.md

Debugger Support for Multi-Module Enclaves
====

This document describes extensions to the
[Debugger Contract](/include/openenclave/internal/debugrt/host.h) to support
debugging enclaves comprised of multiple modules.

Introduction
------------

Currently, enclaves are built as a single module - a special elf executable that
is loaded by OE SDK's enclave loader. Since enclaves are loaded by a special
loader, unlike normal executables and shared libraries that are loaded by the
platform's loader (e.g /lib64/ld-linux-x86-64.so.2), the debugger does not
natively know how to load symbols for enclaves. Therefore, the host side OE SDK
runtime runtime invokes following functions to enable the debugger to locate
and debug an enclave.

- `oe_result_t oe_debug_notify_enclave_created(oe_debug_enclave_t* enclave)`

  This function is called immediately after an enclave has been loaded into
  memory, but before it has been initialized/begun execution. The debugger uses
  the following fields in `oe_debug_enclave_t` structure to load the symbols

>   and mark a region of program memory as being take up by the enclave:

Reviewer: ```suggestion
  and mark a region of program memory as being taken up by the enclave:
```

 ```c

    typedef struct _debug_enclave_t
	{
		const char* path;
		uint64_t path_length;

		const void* base_address;
		uint64_t size;
	} oe_debug_enclave_t;
	
  ```

  The `path` and `path_length` fields tell the debugger the path to the
  enclave's ELF binary. The elf binary is expected to contain the debugging
  information, if any, for the enclave. `oegdb` currently does not support
  debugging enclaves whose debugging information has been stripped and moved to
  a separate .dbg file.

  The `base-address` field tells the debugger the memory address at which the
  enclave has been loaded. The `size` field is not that important since the
  debugger is interested only in the size of the `.text` section which it can
  obtain from the ELF binary. The debugger associates the memory region
  [base-address, base-address+size) with the enclave.

- `oe_result_t oe_debug_notify_enclave_terminated(oe_debug_enclave_t* enclave)`

  This function is called immediately after an enclave has been terminated and

>   it's memory has been deallocated. The debugger unloads the symbols for the enclave

Reviewer: ```suggestion
  its memory has been deallocated. The debugger unloads the symbols for the enclave
```

  and disassociates the memory region from the enclave.

The debugger gains control and responds to the above notifications in the
following ways:

- Linux

  The debugger sets hidden breakpoints in the above notification functions. When
  they are called by the SDK's host-side runtime, the breakpoints are hit and
  the debugger gets control. The debugger then reads the `oe_debug_enclave_t`
  parameter to handle the notification.

- Windows
 

>   Use of breakpoints by the debugger is not recommended. Instead, the Windows

Reviewer: can you say why it's not recommended?  (Recommended by whom?)

Reviewer 2: It's a windows symbol issue. Since PDBs are almost always separate from the DLL it's fairly easy to get into situations where you do not have the symbols available. That makes relying on the debugger setting breakpoints (via symbol data) a lot more fragile. It's much simpler on the debugger side to just receive an event by an exception as that will always work, regardless of whether or not symbols for the oe runtime dlll are available.

Reviewer: @auott But enclaves are ELFs not DLLs so why are PDBs relevant for multi-module enclaves?

Reviewer 2: It depends on where the load notification is coming from - not what is being loaded.
If the load notification comes from a DLL (similar to how the enclave load notification happens) then we have the PDB separation problem. If the notification is coming from inside an enclave, then we don't have the PDB problem.

My understanding was that both those scenarios are possible, though my previous response was about notifications coming from outside an enclave. For the latter case I don't see any issue with relying on symbols.

Author: One thing to keep in mind is that the contract supports Intel SDK's PE enclaves too. 
Instead of relying on debug symbols, would it be OK to rely on exported symbols that are guaranteed to be available even if debug symbols have been stripped off in a binary?
Then we could have both the ELF and PE enclaves export the functions and debugger ought to be able to put breakpoints on these even without debug symbols. However in that case, we'd also need the parameter to passed in a specific register so that the debugger can just read it without requiring debug symbols for the parameter.

Reviewer 2: I think that would be work as well.
The register problem would likely need to be solved anyways as putting a breakpoint based on a symbol would run into the same problem as the breakpoint would likely be at the beginning of the prolog.

  `RaiseException` API is used by the runtime to notify the debugger.
  
  `oe_debug_notify_enclave_created` raises the following exception:
  
  ```c
        ULONG_PTR args[1] = {(ULONG_PTR)enclave};
		RaiseException(
		  OE_DEBUGRT_ENCLAVE_CREATED_EVENT,
		  0, // dwFlags
		  1, // number of args
		  args);
  ```
  
  whereas `oe_debug_notify_enclave_terminated` raises:
  
  ```c
        ULONG_PTR args[1] = {(ULONG_PTR)enclave};
        RaiseException(
          OE_DEBUGRT_ENCLAVE_TERMINATED_EVENT,
          0, // dwFlags
          1, // number of args
          args);
  ```
  
  The Windows debuggers (Visual Studio and WinDbg) consume the
  `OE_DEBUGRT_ENCLAVE_CREATED_EVENT` and `OE_DEBUGRT_ENCLAVE_TERMINATED_EVENT`
  events. Upon receiving the events, the debuggers invoke the appropriate
  infrastructure to enable debugging the module.

Use-cases for Multi-Module Enclaves
-----------------------------------

There are a few implementations of multi-module enclaves based on OE SDK:

- [SGX LKL](https://github.com/lsds/sgx-lkl)

>   This project runs a modified linux kernel within an enclave and supports running

Reviewer: ```suggestion
  This project runs a modified Linux kernel within an enclave and supports running
```

  enclaves comprised of executables and shared libraries. This project supports
  debugging of multi-module enclaves.

>   https://github.com/lsds/sgx-lkl/tree/oe_port/tools/gdb

Reviewer: Here and in 116 and 123, it's not clear what the purpose of the link is.  121-122 at least has some text preceding the link, but what are these?

  This branch of OE SDK runs a library OS within an enclave. It also supports
  enclave comprised of multiple executables and shared libraries.
  https://github.com/openenclave/openenclave/blob/feature.openlibos/debugger/pythonExtension/gdb_sgx_plugin.py

- [feature/dynamic_binding](https://github.com/openenclave/openenclave/tree/feature/dynamic_binding)

  This branch of OE SDK allows loading and running multi-module enclaves with
  the goal of permitting use of FIPS certified crypto modules. It also supports
  debugging multi-module enclaves:
  https://github.com/openenclave/openenclave/commit/254ae94324b619861f58d0110a1affac37e5a2a4?branch=254ae94324b619861f58d0110a1affac37e5a2a4&diff=split

The debugger changes required to support multi-module enclaves are quite similar
in the above three implementations. The following proposal formally incorporates
the necessary debugger changes into the Debugger Contract.

Debugger Contract Extensions to Support Multi-Module Enclaves
-------------------------------------------------------------

The following `oe_debug_module_t` structure captures the information necessary
to debug a module:

```c

>     typedef struct _debug_module_t

Reviewer 2: Similar to the _debug_enclave_t structure, there will need to be a list of the modules that have already been loaded stored somewhere. That way a debugger that attaches can still determine the set of modules without having received the create events for them.

	{
	    // Magic value at start of the structure.

> 		uint64_t magic;

Reviewer: nit: fix indentation to be consistent

		// Structure version.
		uint64_t version;

		// Path of module's ELF binary.
		char* path;
		uint64_t path_length;

		// Memory region where the enclave has been loaded.
		uint64_t base_address;
		uint64_t size;
	} oe_debug_module_t;
	
```

The `path` and `path_length` together specify the path of the module's ELF
binary. The `base_address` and `size` together specify the memory region
where the enclave has been loaded.

The following notification functions are used to notify the debugger about the
loading and unloading of a module.

- `oe_result_t oe_result_t oe_debug_notify_module_loaded(oe_debug_module_t* module)`

   This function is called by the SDK's host-runtime after a module has been
   loaded, but before it has been initialized. A module is initialized by invoking

>    it's constructor functions. It doesn not matter whether the module has been relocated

Reviewer: ```suggestion
   its constructor functions. It does not matter whether the module has been relocated
```

   or not before the debugger is notified about the module.

   In situations where a module may be dynamically loaded by an enclave, as is the case
   when SGX LKL/openlibos launch applications within the enclave, this notification
   function is called by the SDK's enclave-side runtime prior to the initialization
   of the module.

- `oe_result_t oe_debug_notify_module_unloaded(oe_debug_module_t* module)`

   This function is called by the SDK runtimes (host-side or enclave-side), after a
   module has been unloaded.

Debugger Handling of Module Load/Unload Notifications

- Linux

>   Similar to enclave load/unload, the debugger sets internal breakpoints in

Reviewer 3: From the SIG-arch meeting, there was a discussion as to whether this applied to all loaded modules, including the primary enclave app module, and the expectation was that it would exclude the primary enclave module (which would have the equivalent tasks handled as part of the enclave start notification).

The feedback was that it would be nice if the primary enclave could be handled uniformly, with whatever additional tasks specific to enclave start (e.g. TCS handling) being factored out into a separate implementation that is composable with the module load. There was some concern that this could break the existing contract with the windbg, and that would need to be investigated.

Author: While, at first glance, it may seem nicer to treat primary and secondary modules the same, and factor out extra details of the primary module into a separate notification, it does not seem that much valuable to me for the following reasons:
- We are losing data in this abstraction. The fact that a module is a primary module now has to be figured out in another way. 
- An enclave that does not use multiple modules will now require extra notifications to enable debugging where previously it used to do that with fewer notifications.
- The information about the primary module will now be sent via multiple notifications instead of being atomically sent in a single notification. I suspect this could make it difficult for the debugger to correlate information if multiple threads were simultaneously creating enclaves.
- This change is a breaking change that will require a change to all the debuggers (oegdb, VS, WinDbg) to debug single module enclaves. Additionally, it might also require work in Intel SDK to preserve backwards compatibility. It is not clear to me what the bar is for justifying breaking changes in published contracts.

  information for the module.

- Windows
  - Host Side
  
    The SDK host-side runtime uses the `RaiseException` Windows API to
	communicate with the debugger.
	`oe_debug_notify_module_loaded` raises the following exception:
	```c
 
        ULONG_PTR args[1] = {(ULONG_PTR)module};
		RaiseException(
		  OE_DEBUGRT_MODULE_LOADED_EVENT,
		  0, // dwFlags
		  1, // number of args
		  args);
		  
    ```
	
     whereas `oe_debug_notify_enclave_terminated` raises:
     ```c
   
        ULONG_PTR args[1] = {(ULONG_PTR)module};
        RaiseException(
          OE_DEBUGRT_MODULE_UNLOADED_EVENT,
          0, // dwFlags
          1, // number of args
          args);
		  
	```
	
  - Enclave side
    
	The enclave side runtime cannot call Windows API functions. Instead, it
	relies on `int 3` exception to communicate with the debugger. The `int 3`
	instruction is preceeded by special byte sequence `0x0edeb0gca11bac00` to
	allow the debugger to distinguish between these special `int 3` exceptions
	and any other `int 3` exceptions that may occur.
	
	`oe_debug_notify_module_loaded` uses the following instruction pattern:
	```asm

>        mov module, %rax

Reviewer: The same ELF file should run on both Windows and Linux, so you have to insert this at compile time without know whether it will run on Windows or Linux or both.

Reviewer 2: In the windows case we should be able to rely on symbols for the enclave portion (since they should be embedded in the ELF file). That way both the Linux and Windows in-enclave load can be handled the same way via internal breakpoints.

	   mov $1, %rbx     // 1 => ModuleLoaded

> 	   jmp do_int3

Reviewer 3: From the SIG-arch meeting: @binxing had a suggestion about using `nop` and having the debugger overwrite it with the `int 3` as needed, which would make the code path uniform on the enclave side regardless of whether the enclave is under debug. There was an open question as to whether the EDBGWR instruction would fault if the page itself is RX-only.

Reviewer 4: EDBGWR ignores the EPCM RWX permissions (it does require that the enclave is loaded in debug mode).  Thus, the OS interfaces (WriteProcessMemory for Windows and ptrace for Linux) should work the same for when a debugger writes to non-enclave memory.

Author: Austen and I discussed this. While this is an elegant proposal, it seems like it is quite close to what the debugger does underneath the hood for a normal breakpoint. Is there any benefit this approach would have in comparison to a hidden breakpoint?

	   .bytes 0x0edeb0gca11bac00
	do_int3:
	   int $3
	   
	```
	whereas `oe_debug_notify_module_unloaded` uses the following instruction
	pattern:
	```asm
	   
	   mov module, %rax
	   mov $2, %rbx     // 2 => ModuleUnloaded
	   jmp do_int3
	   .bytes 0x0edeb0gca11bac00
	do_int3:
	   int $3
	   
	```

    When the debugger receives and `int 3` exception, it first checks the bytes
	immediately preceeding the program counter. If the bytes match the special
	sequence, then the debugger reads `rax` register to obtain the pointer to
	`oe_debug_module_t` structure, and `rbx` register to determine whether
	the module was loaded or unloaded.

The Debugger Contract version is incremented to `2` to reflect the addition of
the notification functions.

> The host-side runtime will set [oe_debugger_contract_version](https://github.com/openenclave/openenclave/blob/6f94547dc920d70eb50ec8ca95ab65a17597364b/include/openenclave/internal/debugrt/host.h#L131)

Reviewer 2: Note: I don't believe the Windows debuggers will currently do this, though this could definitely be a future behavior.

  It is uses a quite similar strategy to the above implementations:
  https://github.com/openenclave/openenclave/commit/254ae94324b619861f58d0110a1affac37e5a2a4?branch=254ae94324b619861f58d0110a1affac37e5a2a4&diff=split

The debugger changes required to support multi-module enclaves are quite similar
in the above three implementations. The following proposal formally incorporates
the necessary debugger changes into the Debugger Contract.

Debugger Contract Extensions to Support Multi-Module Enclaves
-------------------------------------------------------------

The following `oe_debug_module_t` structure captures the information necessary
to debug a module:

```c

    typedef struct _debug_module_t
    {
        // Magic value at start of the structure.
        uint64_t magic;
        // Structure version.
        uint64_t version;

        // Path of module's ELF binary.
        char* path;
        uint64_t path_length;

        // Memory region where the enclave has been loaded.
        uint64_t base_address;
        uint64_t size;

        // Next item in list of module.
        struct _debug_module_t* next;

>     } oe_debug_module_t;

Reviewer 4: If an oe_debug_module_t structure is added to the linked list of modules before the oe_debug_notify_module_loaded() function is called and then removed after the oe_debug_notify_module_unloaded() function is called, then a debugger that attaches to a running process could get an event on a module that it already sees as loaded, or miss an event on a module that has been unloaded.

Author: Thanks for pointing this out Mark. As always, I appreciate your insights.

The following three operations are themselves not atomic, and also cannot be performed together atomically:
- Check if the debugger has been attached or not
- Notify the debugger
- Add an entry to the module list

Due to the lack of atomicity, the ordering of the operations can result in different edge-cases.
(The "check if debugger has been attached or not" is ignored from the scenarios below since debugger may immediately attach/detach after the check has been performed; making the check unreliable).

In one ordering, the debugger may get double notification:
```
    1. Runtime adds module to list
        <---- Debugger attaches to the program and processes the module
    2. Runtime notifies the debugger about the module <---- Debugger will have to avoid processing the module again
    
```

In another approach, the debugger may lose track of modules
```
    1. Runtime notifies the debugger  <-- notification ignored since debugger not attached
      <------ Debugger attaches to the program, processes the list and finds no module
    2. Runtime adds the module to the list  <--- list has already been processed; module is "lost"
```

I don't know if there is a way to solve the edge-case elegantly. We could go with the first ordering above and require that the debugger be prepared for double-notifications for parsed modules upon attach.

Note: The same ordering problem exists for the list of enclaves. I'd have to dig up my notes/mails, but IIRC, this edge-case was determined not too cost-effective to address given the timing constraints of the debugger contract.

Author: I discussed with @auott, the strategy that was used for handling the list of enclaves what that the debugger ought to handle potential double-notification during attach.

is loaded by OE SDK's enclave loader. Since enclaves are loaded by a special
loader, unlike normal executables and shared libraries that are loaded by the
platform's loader (e.g /lib64/ld-linux-x86-64.so.2), the debugger does not
natively know how to load symbols for enclaves. Therefore, the host side OE SDK
runtime runtime invokes following functions to enable the debugger to locate
and debug an enclave.

- `oe_result_t oe_debug_notify_enclave_created(oe_debug_enclave_t* enclave)`

  This function is called immediately after an enclave has been loaded into
  memory, but before it has been initialized/begun execution. The debugger uses
  the  `oe_debug_enclave_t` structure to load the symbols and mark a region of
  program memory as being taken up by the enclave:

 ```c

    typedef struct _debug_enclave_t
	{
        uint64_t magic;

        uint64_t version;

        // Pointer to next enclave
		// See "Attaching the debugger" section below
        struct _debug_enclave_t* next;

        const char* path;
        uint64_t path_length;

        const void* base_address;
        uint64_t size;

        // Enclave's TCS information.
        struct _sgx_tcs** tcs_array;

>         uint64_t num_tcs;

Reviewer: ```suggestion
        uint64_t tcs_count;
```
Per OE style guidelines

	for PE enclaves. Can we require that the enclave side notification functions
	are exported so that the debuggers are guaranteed to always find them?

The Debugger Contract version is incremented to `2` to reflect the addition of
the notification functions.

The host-side runtime will set [oe_debugger_contract_version](https://github.com/openenclave/openenclave/blob/6f94547dc920d70eb50ec8ca95ab65a17597364b/include/openenclave/internal/debugrt/host.h#L131)
to `2` during program startup. If an older debugger is used to debug a
multi-module enclave, then it will see that the contract version is `2`, and
advise the user to upgrade the debugger.

### Attaching the debugger

In order to enable the debugger to discover modules in the scenario where the
debugger is attached after the modules have been loaded, a list of modules
is maintained in the `oe_debug_enclave_t` data structure.

typedef struct _debug_enclave_t
{
    uint64_t magic;

    uint64_t version;

    struct _debug_enclave_t* next;

    const char* path;
    uint64_t path_length;

    const void* base_address;
    uint64_t size;

    struct _sgx_tcs** tcs_array;

>     uint64_t num_tcs;

Reviewer: ```suggestion
    uint64_t tcs_count;
```
or tcs_array_size

    uint64_t flags;

	// New field added to the end

> 	oe_debug_module_t secondary_modules;

Reviewer: nit: fix indentation
